decisions/DR009-Batch-method-result-types.md

```diff
@@ -0,0 +1,296 @@
+# DR009 Batch method result types
+
+- **Status:** Proposed
+- **Impact:** High
+- **Driver:** @feltech
+- **Approver:** @foundrytom @feltech
+- **Contributors:** @foundrytom @feltech
+- **Outcome:** Batch methods will adopt a per-result callback mechanism.
+
+## Background
+
+OpenAssetIO employs a batch-first design. In that, all data centric
+methods take multiple inputs, and produce multiple outputs. This allows
+a `ManagerPlugin` to optimize queries to potentially high-latency
+back-end services.
+
+Many methods may have per-value error states, such that the output of
+these methods is never of a uniform type. For example, `resolve` for any
+given entity reference may provide a `TraitsData` instance, or an error.
+An error for a specific value should not error the entire batch.
+
+This raises the question of how to best express the resulting data across
+the variety of languages supported by OpenAssetIO.
+
+## Relevant data
+
+OpenAssetIO core is currently exposed through C, C++ and Python. The C
+API serves as a binding point for other languages.
+
+A design mantra is that the implementation in any given language should
+be as idiomatic as possible. An exception to which is that Python and
+C++ are often used in the same breath within M&E DCC applications,
+keeping those two languages as close to each other as possible has many
+benefits.
+
+The performance of API methods is important, as many scenarios
+(particularly `resolve`) have a high query volume. Syntax alone is not
+enough of a justification for any particular design choice.
+
+For the scope of this document, it is assumed that the input to any of
+these methods will be an array, regardless of the choices here.
+
+Input options are not explored here, as the sole purpose of the
+batch-centric pattern is to optimise queries, which requires all input
+data to be submitted as a single request to back-end services.
+
+The OpenAssetIO API is entirely synchronous. There have been discussions
+around the merit of a future-based variant, but so far all explored
+scenarios have favoured a simple blocking API. This is a topic that can
+be revisited at a later date should the need arise.
+
+Note also that by design, all API methods (aside from initialization)
+are thread-safe and may be called concurrently from any number
+of threads.
+
+## Options considered
+
+### Option 1
+
+Return an array of 'result or error' objects for each input value.
+
+#### Pros
+
+- More familiar in target languages.
+- Facilitates data-centric functional programming without caller-side
+  wrapping.
+- Data-centric result is easier to propagate across process/host
+  boundaries.
+
+#### Cons
+
+- The result type is complex and requires significant boilerplate
+  on the host side to un-pack.
+- Requires additional allocations/copies of the result container.
+
+### Option 2
+
+Provide success/error callbacks to be invoked for each input value.
+
+The manager guarantees to invoke one of the callbacks for each input
```

Comment:
[question] Is is worth a note on exceptions, or is that just muddying the waters?

Reply:
We def need to address this (especially in terms of exceptions thrown during the callback), can we save this for the full design once the high-level choice of callback or not is agreed?